test/integration/banListTest.ts

```diff
@@ -0,0 +1,80 @@
+ * @param reason A reason for the rule e.g. 'Wouldn't stop posting spam links'
+ * @returns The event id of the newly created policy rule.
+ */
+async function createPolicyRule(client: MatrixClient, policyRoomId: string, policyType: string, entity: string, reason: string) {
+    return await client.sendStateEvent(policyRoomId, policyType, `rule:${entity}`, {
+        entity,
+        reason,
+        recommendation: 'm.ban'
+    });
+}
+
+describe("Test: Updating the BanList", function () {
+    it("Calculates what has changed correctly.", async function () {
+        this.timeout(10000);
+        const mjolnir = config.RUNTIME.client!
+        const moderator = await newTestUser(false, "moderator");
+        const banListId = await mjolnir.createRoom({ invite: [await moderator.getUserId()]});
+        const banList = new BanList(banListId, banListId, mjolnir);
+        mjolnir.setUserPowerLevel(await moderator.getUserId(), banListId, 100);
+
+        // Test adding a new rule
+        await createPolicyRule(mjolnir, banListId, RULE_USER, '@added:localhost:9999', '');
+        let changes: ListRuleChange[] = await banList.updateList();
+        assert.equal(changes.length, 1, 'There should only be one change');
+        assert.equal(changes[0].changeType, ChangeType.Added);
+        assert.equal(changes[0].sender, await mjolnir.getUserId());
+
+        // Test modifiying a rule
+        let originalEventId = await createPolicyRule(mjolnir, banListId, RULE_USER, '@modified:localhost:9999', '');
+        await banList.updateList();
+        let nextEventId = await createPolicyRule(mjolnir, banListId, RULE_USER, '@modified:localhost:9999', 'modified reason');
+        changes = await banList.updateList();
+        assert.equal(changes.length, 1);
+        assert.equal(changes[0].changeType, ChangeType.Modified);
+        assert.equal(changes[0].previousState['event_id'], originalEventId, 'There should be a previous state event for a modified rule');
+        assert.equal(changes[0].event['event_id'], nextEventId);
+
+        // Test redacting a rule
+        const redactThis = await createPolicyRule(mjolnir, banListId, RULE_USER, '@redacted:localhost:9999', '');
+        await banList.updateList();
+        await mjolnir.redactEvent(banListId, redactThis);
+        changes = await banList.updateList();
+        assert.equal(changes.length, 1);
+        assert.equal(changes[0].changeType, ChangeType.Removed);
+        assert.equal(changes[0].event['event_id'], redactThis, 'Should show the new version of the event with redacted content');
+        assert.equal(Object.keys(changes[0].event['content']).length, 0, 'Should show the new version of the event with redacted content');
+        assert.notEqual(Object.keys(changes[0].previousState['content']), 0, 'Should have a copy of the unredacted state');
+        assert.notEqual(changes[0].rule, undefined, 'The previous rule should be present');
+
+        // Test soft redaction of a rule
```

Comment:
Out of curiosity, what's "soft redaction"?

Reply:
I just made it up, we a find a better name if needed but it's overwriting the room state for that combination of state key and state type with a new event that has empty content.

As a semi related aside (but an important one). This is how Mjolnir currently marks the rules as 'deleted' this way rather than sending a redaction event. One reason for doing this is so that Mjolnir only has to listen to the policy event types rather than every single redaction (and then attempt to check every event that the redaction refers to..) in order to make sure lists are kept up to date in soft real time.